frontend/utils/distribution/monthly-event-parser.js

```diff
@@ -0,0 +1,241 @@
+    groupMembers = groupMembers.filter((v) => { return v.name !== event.data.name })
+  }
+
+  // Loop through the events, pro-rating each user's monthly pledges/needs:
+  distributionEvents.forEach((event) => {
+    if (event.type === 'startCycleEvent') {
+      handleCycleEvent(event)
+    } else if (event.type === 'haveNeedEvent') {
+      handleIncomeEvent(event)
+    } else if (event.type === 'paymentEvent') {
+      handlePaymentEvent(event)
+    } else if (event.type === 'userExitsGroupEvent') {
+      handleExitEvent(event)
+    }
+  })
+
+  const lastWhen = distributionEvents[distributionEvents.length - 1].data.when
+
+  const artificialEnd = {
+    type: 'startCycleEvent',
+    data: {
+      when: lastWhen,
+      monthlyDistribution: [] // List to be populated later, by the events-parser
+    }
+  }
+  handleCycleEvent(artificialEnd, true)
+
+  // "Overpayments sometimes occur *internally* as a result of people leaving, joining, and (re-)setting income.
+  // Our task is to redistribute the overpayments back into the current late payments so nobody in need is asked to pay.
+  let overPayments = JSON.parse(JSON.stringify(lastStartCycleEvent.data.monthlyDistribution)).filter((p) => {
+    return p.amount < 0
+  }).map((p) => {
+    p.amount = Math.abs(p.amount)
+    p.total = Math.abs(p.total)
+    return p
+  })
+
+  startCycleEvent.data.monthlyDistribution = redistributeOverToLatePayments(overPayments, startCycleEvent.data.monthlyDistribution)
+
+  if (!adjusted) {
+    startCycleEvent.data.monthlyDistribution = reduceDistribution(addDistributions(startCycleEvent.data.completedPayments, startCycleEvent.data.monthlyDistribution))
+  }
+
+  overPayments = JSON.parse(JSON.stringify(startCycleEvent.data.monthlyDistribution)).filter((p) => {
+    return p.amount < 0
+  }).map((p) => {
+    p.amount = Math.abs(p.amount)
+    p.total = Math.abs(p.total)
+    return p
+  })
+
+  lastStartCycleEvent.data.monthlyDistribution = redistributeOverToLatePayments(overPayments, lastStartCycleEvent.data.monthlyDistribution)
+  // Unadjust last
+  if (!adjusted) {
+    lastStartCycleEvent.data.monthlyDistribution = reduceDistribution(addDistributions(lastStartCycleEvent.data.completedPayments, lastStartCycleEvent.data.monthlyDistribution))
+  }
+
+  lastStartCycleEvent.data.monthlyDistribution.forEach((payment) => {
+    payment.amount = Math.min(payment.amount, payment.total)
+    payment.isLate = true
+    payment.dueOn = dateToMonthstamp(lastDayOfMonth(dateFromMonthstamp(prevMonthstamp(dateToMonthstamp(new Date(lastWhen))))))
+  })
+  startCycleEvent.data.monthlyDistribution.forEach((payment) => {
+    payment.amount = Math.min(payment.amount, payment.total)
+    payment.isLate = false
+    payment.dueOn = dateToMonthstamp(lastDayOfMonth(dateFromMonthstamp(dateToMonthstamp(new Date(lastWhen)))))
+  })
+
+  const previousTwoCycles = [
+    lastStartCycleEvent.data.monthlyDistribution,
+    startCycleEvent.data.monthlyDistribution
+  ].flat()
```

Comment:
I would prefer using a pair of `...` operators here rather than `.flat()`, although that's no big deal so feel free to ignore this if you prefer `.flat()`:
```js
const previousTwoCycles = [
  ...lastStartCycleEvent.data.monthlyDistribution,
  ...startCycleEvent.data.monthlyDistribution
]
```

Reply:
Thanks for the comment @snowteamer. I marked this as resolved because I have now switched to using your way, in most cases.